PyTorch ThroughputBenchmark

salexspb opened the pull request:
Summary:
This is useful for measuring inference performance of your
models. This is a very basic benchmark for now. We don't support
batching on the benchmark side, no inter and intra op parallelizm is
supported yet, just caller based parallelizm.

Main phylosophy here is that user should be able to provide inputs
from python and just stack them within the benchmark. API should be
exactly the same as passing inputs to module.forward.

Differential Revision: D15435461

soumith reviewed (changes requested):
this should be a separate binary / utility, outside of csrc -- from what I see it doesn't need to hook into any of the internals in any way.
Probably going into the `benchmark` folder as a separate utility?

dzhulgakov commented:
pytorch/benchmarks feels like indeed a better place. You can make it a separate binary/library and link against libtorch.so

The reasoning is that it's more of a scaffolding around pytorch that doesn't require changes in the core (like e.g. autograd profiler). So putting it separately is cleaner. Even if we were put it in the same build, csrc is probably the wrong place to put it. cc @gchanan for input too.

salexspb commented:
@dzhulgakov , @soumith , my motivation is that I would like any researcher easily see perf of their model in an inference like setup. It is not specific to a model, it is a generic benchmarking utility. 

So yes, I would like it to be part of the same build. Then one doesn't need to recompile anything after they developed a model and can just call this utility. How I see it, it should work this way both in OSS and internally. 

No issues pytorch/benchmarks folder specifically, but would it work the way I described above?

dzhulgakov commented on test/test_throughput_benchmark.py:
```diff
@@ -0,0 +1,58 @@
+        config.num_calling_threads = 4
+        config.num_warmup_iters = 100
+        config.num_iters = 1000
+        bench = ThroughputBenchmark(module._c, config)
```
you probably want a python wrapper that hides `_c` part in it

salexspb replied:
yeah, python wrapper can  make other things easier in the future as well

dzhulgakov commented on test/test_throughput_benchmark.py:
```diff
@@ -0,0 +1,58 @@
+        with NamedTemporaryFile() as f:
+            torch.jit.save(module, f.name)
+
+        config = BenchmarkConfig()
```
why not just make them arguments of ThroughputBenchmark function?

or even better - make them arguments of .benchmark() method

salexspb replied:
Oh, named arguments, nice! :)

salexspb replied:
@dzhulgakov , actually having a config struct should be useful going forward. One can supply million of various knobs. Function signature will become really bad on the c++ side if we keep pushing those as arguments to a function. 

For python we can have a wrapper which under the hood creates the config structure and passes it over to c++.

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,125 @@
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
+        VLOG(2) << "Completed forward pass on thread " << thread_id;
+        ++num_forwards;
```
you could be even more precise by putting `++num_forwards >= config.num_iters` here and measuring time directly

salexspb replied:
Yes, I can make this one atomic operation. What do you mean by "measuring time directly"?

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,125 @@
+      while (!shutdown) {
+        VLOG(2) << "Starting forward pass on thread " << thread_id;
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
```
it can potentially go out of bounds - there's no guarantee that the main thread will stop things in time

salexspb replied:
oops. Yeah, probably just better to make number of total iterations precise

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,125 @@
+  float total_time_ms = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      end_time - start_time).count() / 1000.0 / 1000.0;
+  stats.set_latency_avg_ms(total_time_ms / num_forwards);
+  stats.set_qps(num_forwards / total_time_ms * 1000.0);
```
iters per sec? qps is kinds of internal slang

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.h:
```diff
@@ -0,0 +1,164 @@
+ */
+class ExecutionStats {
+public:
+  float latency_avg_ms() const {
```
I think pybind should just let you bind fields directly without getters/setters - so this entire thing can be a struct (https://pybind11.readthedocs.io/en/stable/classes.html#instance-and-static-fields - def_readonly, def_readwrite)

dzhulgakov reviewed (commented):
Hm, if we want to expose it in the main build, I guess it doesn't make sense to build it as a separate python extension.

The question becomes whether it cuts the bar for general usefulness (similarly to autograd profiler) - @soumith @gchanan . But it that case we probably want to put it into some sub-namespace (torch.profiler? torch.bechmark? autograd.profiler name is a bit unfortunate)

salexspb commented:
Putting things into a different python namespace makes sense to me, putting things into jit was rather to cause this discussion and figure out what is the proper place :) 

@soumith , @zdevito , @gchanan , given the goal I described above (of making this easily available to any researcher / ml practitioner so they can test their model throughput without additional builds), should I just go ahead and create a new python namespace like torch.benchmark? I assume, in this case, the c++ code stays in the same place, but please let me know if there is a better place for it. 

autograd.profiler should likely also live there :)

zdevito commented:
We should probably provide a torch.profiler package as part of the main build or in an easily installable package similar to torchvision. First, this is where autograd profiler should live. It has long since become more generic than just profiling autograd. We can make torch.autograd.profiler alias torch.profiler for now. Second, a generic harness for timing the throughput and latency of a model until appropriate multi-core load is a useful enough tool that it seems to make sense going into torch.profiler. Note that this will enforce a higher barrier of quality on it than if it were for our own use only, but that is probably a good thing because we to make it easier to accurately benchmark our models anyway. It can serve as a way to capture best practices for our external users. Example: JIT requires warm up (at least one iteration). I don't expect users to know this, so if they write their own benchmark harness then they often report slower numbers.

zdevito commented on test/test_throughput_benchmark.py:
```diff
@@ -0,0 +1,58 @@
+        config.num_calling_threads = 4
+        config.num_warmup_iters = 100
+        config.num_iters = 1000
+        bench = ThroughputBenchmark(module._c, config)
```
_c is private, please do not rely on it outside the implementation of ScriptModule

zdevito reviewed (commented):
Seems like a good first pass. If this is going to go into PyTorch proper and have python bindings, it needs to work with nn.Module not just ScriptModules.

soumith commented:
There is precedence (and good design) for this. Take a look at `torch.utils.bottleneck`: https://pytorch.org/docs/stable/bottleneck.html

It was a tool that was aimed to tell users in a one-touch way what part of their code is a bottleneck -- without the user understanding how to use the profiler, or plug profiler code deeper into their codebase.

soumith commented on test/test_throughput_benchmark.py:
```diff
@@ -0,0 +1,58 @@
+        for input in inputs:
+            bench.add_input(input)
+        for i in range(NUM_INPUTS):
+            module_result = module.forward(inputs[i])
```
`module(inputs[i])`, not `module.forward`

soumith replied:
this still needs to be fixed

dzhulgakov replied:
address this?

soumith commented:
similarly, we can probably aim for `torch.utils.benchmark`, but I think the API needs to be cleaned up a bit. For example `config` can be removed in favor of just using args / kwargs (if you choose, you can pass a dict in as kwargs with `throughput(**config)` or something

soumith commented:
the other thing I can think of that the API needs to be improved upon is beyond CPU throughput.
It's pretty common to do GPU-inference. If so, what exactly are the tuning parameters there (like for CPU it's number of threads etc.)

salexspb commented:
So here is a new version. I think, I addressed all the API questions except the following:

1. @zdevito  suggestion to support nn.Module. Do you suggest we should just take an nn.Module and trace it? Because IMO running python code through throughput inference benchmark doesn't make much sense, people don't really deploy python. At least this is not a big use-case which I would like to not be blocked on. If you suggest just trying to trace the module, it should be a few lines change.

2. @soumith 's comment about supporting GPU inference. I think this is a good goal. API wise calling threads should stay there . In majority of the cases calling threads are CPU threads which offload to an accelerator card. Speaking of which - we actually would want to support other hardware as well. I think, we may provide additional parameter here to the benchmark which device to run model on. With future PEP integration we hopefully will allow people to submit remote jobs from an iPython notebook and run this whole thing on a remote host with hardware they want to test against. How cool would be that ? :)  

I also didn't move the module location (it is currently under pytorch.jit), I intend to more it to torch.profiler as soon as we agree on this step for sure :) 

Update summary:

1. nicer api for callling benchmark from Python. Now benchmkar() method takes in key word arguments which  I convert to BechmarkConfig struct later in the python helper. I would like to keep the struct for C++ API as C++ is not good about key word arguments and I don't want benchmark() c++ method to become a mess. 

2. hidden model._c call

3. more documentation

4. @dzhulgakov 's suggestions about getters and setters

5. @dzhulgakov 's suggestions on ExecutionStats field names

6. fixed out of bound issue.

kimishpatel commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,124 @@
+      }
+      LOG(INFO) << "Initialized thread " << thread_id;
+      ++initialized;
+      cv.notify_all();
```
Maybe I did not understand but is this notify_all needed?

salexspb commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,124 @@
+
+  {
+    std::unique_lock<std::mutex> g(m);
+    cv.wait(g, [&]() { return initialized == config.num_worker_threads; });
```
@kimishpatel , the notify_all you asked about triggers this cv.wait. If we see all threads initialized, we will set start = true.

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,124 @@
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), inputs_.size());
```
why .size()?

dzhulgakov commented on torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,124 @@
+  LOG(INFO) << "Finished benchmark";
+
+  ExecutionStats stats;
+  float total_time_ms = std::chrono::duration_cast<std::chrono::nanoseconds>(
```
you could probably cast to something like `duration<std::chrono::milliseconds, float>` but I'm not sure how exactly it works

dzhulgakov commented on torch/csrc/throughput_benchmark/init.cpp:
```diff
@@ -0,0 +1,63 @@
+      .def_readwrite("num_warmup_iters", &BenchmarkConfig::num_warmup_iters)
+      .def_readwrite("num_iters", &BenchmarkConfig::num_iters);
+
+  py::class_<ExecutionStats>(m, "ExecutionStats")
```
BenchmarkExecutionStats? I'm worried that we bind too generic name here

dzhulgakov replied:
address this?

dzhulgakov commented on torch/csrc/throughput_benchmark/init.cpp:
```diff
@@ -0,0 +1,63 @@
+      .def(py::init<std::shared_ptr<jit::script::Module>>())
+      .def(
+          "add_input",
+          [](py::args args, py::kwargs kwargs) {
```
you can make the first argument ThroughputBenchmark& directly - from docs:

> You may also use just one or the other, and may combine these with other arguments as long as the py::args and py::kwargs arguments are the last arguments accepted by the function.

dzhulgakov replied:
did it work?

salexspb replied:
yep

dzhulgakov reviewed (commented):
Looks nice, I'd defer to @gchanan on naming. My guess would be `torch.profiler.ThroughputBenchmark`

salexspb commented:
spoke with @zdevito , going to add support for nn.Module in order to be more generic.

soumith commented:
the final location should be in `torch.utils`, I'm pretty sure it should be `torch.utils.throughput_benchmark`, unless there is somehow a natural place for it in `torch.autograd.profiler`, which I think there isn't.

I also think `throughput_benchmark` is a long name, so I wonder if `throughput` has alternatives, but I dont care about bikeshedding it too much.

dzhulgakov commented on caffe2/CMakeLists.txt:
```diff
@@ -442,6 +442,8 @@ if (NOT INTERN_BUILD_MOBILE)
     ${TORCH_SRC_DIR}/csrc/jit/fuser/executor.cpp
     ${TORCH_SRC_DIR}/csrc/jit/fuser/codegen.cpp
     ${TORCH_SRC_DIR}/csrc/jit/fuser/fallback.cpp
+    ${TORCH_SRC_DIR}/csrc/throughput_benchmark/init.cpp
```
nit: PT folder structure is relatively flat, it feels like a wrong top-level directory. How about calling it csrc/profiling or csrc/perftools ?

dzhulgakov commented on torch/jit/__init__.py:
```diff
@@ -1974,3 +1974,71 @@ def _graph_for(self, *args, **kwargs):
+    '''
+
+    def __init__(self, module):
+        if isinstance(module, torch.jit.ScriptModule):
```
hm, not sure whether there's a better way, but I guess it's ok

salexspb commented:
@dzhulgakov , sorry for the churn, I addressed XQ's feedback on the internal diff wrt CVs usage, the python module part needs some update.

salexspb commented:
@soumith , @dzhulgakov , @zdevito , the new version is ready. I made it work with nn.Module and addressed the comments (including the ones @zheng-xq left internally). 
Main changes:
1. template BenchmarkHelper in order to support different model representations
2. moved the code according the the discussion above. At least how I interpreted it :) i.e. csrc/perf_tools for c++ part and utils/throughput_benchmark for python part.

salexspb commented:
OMG, looks like builds are passing! :) 
Used _dr.ci tool and looks like all the failures are not related.

Kindly asking for a review :)

soumith commented on tools/build_variables.py:
```diff
@@ -233,6 +233,8 @@ def add_torch_libs():
         "torch/csrc/serialization.cpp",
         "torch/csrc/tensor/python_tensor.cpp",
+        "torch/csrc/perf_tools/init.cpp",
+        "torch/csrc/perf_tools/ThroughputBenchmark.cpp",
```
throughput_benchmark.cpp (sorry for the nit, but all other files in csrc seem to be consistent that way)

soumith commented on torch/CMakeLists.txt:
```diff
@@ -82,6 +82,8 @@ set(TORCH_PYTHON_SRCS
     ${TORCH_SRC_DIR}/csrc/multiprocessing/init.cpp
     ${TORCH_SRC_DIR}/csrc/onnx/init.cpp
+    ${TORCH_SRC_DIR}/csrc/perf_tools/init.cpp
+    ${TORCH_SRC_DIR}/csrc/perf_tools/ThroughputBenchmark.cpp
```
more bike-shedding, `csrc/utils/throughput_benchmark.cpp` to make it consistent with `torch.utils` on the Python side? (more of a soft question)

salexspb replied:
@soumith , sure, consistency is good and we should keep things consistent. 
One issue with "utils" name is that it becomes a bug of stuff. I guess it can be a smaller issue with PyTorch given extensive reviews we have. But in Facebook such general things usually end up being a mess.

What do you think, do you prefer to keep torch.utils.ThroughputBenchmark? Another option I see is to create a new module called torch.perf_tools. This also can motivate people to contribute more to the space :)

soumith replied:
i think we are and will be diligent about code dumping. definitely dont want it to become a mess.

once we actually think through how to extend this utility to be a more general throughput / serving simulator (for example serving GPU or other accelerators, what the API would look like to generalize over -- what would serving look like via multiprocessing), i think we should definitely open up a `torch.profiling` namespace. I am just wary of opening up the namespace now with this single utility in it's current form.

salexspb replied:
Fair enough, I like this approach!

soumith commented on torch/csrc/perf_tools/ThroughputBenchmark-inl.h:
```diff
@@ -0,0 +1,143 @@
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), "Please provide benchmark inptus");
```
"Please provide benchmark inputs" is not a helpful error message, i think.
Also say what was provided. Maybe: "inputs was expected to be non-zero number of elements, but got inputs to be zero elements."

salexspb replied:
This message will fire in a case user didn't call num_input at all. This is the only way to provide inputs, you can not explicitly pass an empty array with the current API. 
I can change this message to talk about add_input() call not being made.

soumith commented on torch/csrc/perf_tools/ThroughputBenchmark-inl.h:
```diff
@@ -0,0 +1,143 @@
+    TORCH_CHECK(!inputs_.empty(), "Please provide benchmark inptus");
+    std::uniform_int_distribution<int> dist(0, inputs_.size() - 1);
+
+    for (int thread_id = 0; thread_id < config.num_calling_threads;
```
looking at this, I'm curious if you want to set `FTZ` as an option or not in benchmark inputs. ill-conditioned inputs might definitely come in at runtime, and have really adverse perf impact without FTZ flag.

salexspb replied:
@soumith, Hmmm. I don't think we ever did it in C2 production. 
One approach overal for providing knobs here I am thinking about - we will evolve things as we go and validate them on real use-cases. Otherwise I don't have a good use-case to validate effects of enabling FTZ as of right now. 

We also would need some sort of evaluation + benchmark setting in order to measure this. 

Thanks for bringing this up though! I added this to our benchmarking planning document.

I think in order to help to discover this opportunity for people, we can add it to the sweep tool. We can automatically find that FTZ is helpful and thus provide a recommendation. Then use can do their own evaluation with FTZ enabled.

soumith commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,88 @@
+    to support inter and intra-op parallelism as well as multiple models
+    running in a single process.
+
+    Please note that even though nn.Module is support, it might incur an overhead
```
"even though nn.Module is support" -> "even though nn.Module is supported"

soumith commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,88 @@
+
+        for input in inputs:
+            # Both args and kwargs work, same as any PyTorch Module / ScriptModule
+            bench.add_input(input[0], x2=input[1])
```
i'm not sure from reading the doc in this file on what `add_input` does. Does it add a set of inputs, and if you call `add_input` multiple times with different inputs it will pick one input set at random for a benchmark iteration?

salexspb replied:
yes, it will pick up a random input at each iteration (almost, I pre-generate  random indexes). Will update this

soumith commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,88 @@
+
+    def add_input(self, *args, **kwargs):
+        '''
+        Store a single input to a module into the benchmark memory
```
improve doc for this method

soumith commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,88 @@
+        num_calling_threads Calling threads are those threads that are calling
+        into a module in parallel.
+
+        num_warmup_iters Warmup iters are used to make sure we run a module
```
the docstrings can be improved by following the Args:: format, with the added benefit that they'll be rendered beautifully in the final docs:
https://github.com/pytorch/pytorch/blob/master/torch/_tensor_docs.py#L584-L588
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.copy_

soumith commented:
overall, pretty close to being merged :)

dzhulgakov commented on torch/csrc/perf_tools/ThroughputBenchmark.h:
```diff
@@ -0,0 +1,169 @@
+
+struct TORCH_API ModuleInput {
+  ModuleInput(ModuleInput&& other) = default;
+  ModuleInput(const ModuleInput&) = default;
```
maybe just make it move-only? then it's "safer" to manipulate it outside of GIL

also, py::args is a py::tuple and is a py::object. Move constructor doesn't touch refcount: https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L238 - you should be able to remove a lot of GIL-iness

salexspb replied:
hmmmm. As I mentioned above, I got pretty unlucky with GIL around runOnce(). But you are right, it should not touch it. Very strange, will check it out.

salexspb replied:
OMG, all this time it was destructor of the result, not the move itself! So my very first version I did here got the GIL right, I just should have not returned anything here.

dzhulgakov commented on torch/csrc/perf_tools/ThroughputBenchmark.h:
```diff
@@ -0,0 +1,169 @@
+
+  bool initialized() const { return initialized_; }
+
+  std::vector<Input> inputs_;
```
here's a trick question, destruction of ModuleInput requires to be under GIL (I suppose). How can you achieve that? Do we assume that destructor of BenchmarkHelper itself is called under GIL?

salexspb replied:
I would assume so. When python decides to drop benchmark object, its ref count will go to zero from the python thread and it will be under GIL this way. What do you think?

dzhulgakov replied:
yeah, should be ok. I'd add a comment about it as it's non-obvious

dzhulgakov reviewed (changes requested):
Getting there!

salexspb commented on torch/csrc/perf_tools/ThroughputBenchmark-inl.h:
```diff
@@ -0,0 +1,143 @@
+      // performs required warmeup iterations before we start measuring
+      for (auto j = 0; j < config.num_warmup_iters; ++j) {
+        if (std::is_same<Input, ModuleInput>::value) {
+          AutoGIL gil_guard;
```
@dzhulgakov , I was not able to get this code working even with custom move constructor injecting GIL :'( So ended up just GILing here. I will try mutable reference, it will most likely work, thanks for the idea!

dzhulgakov replied:
How does it fail if one std::move() for py::object? I find it surprising

salexspb replied:
OMG, all this time it was destructor of the result, not the move itself! So my very first version I did here got the GIL right, I just should have not returned anything here.

salexspb commented on torch/csrc/perf_tools/ThroughputBenchmark.cpp:
```diff
@@ -0,0 +1,114 @@
+    AutoNoGIL no_gil_guard;
+    c10::IValue result =
+        script_module_.runOnce(std::move(args), std::move(kwargs));
+    return jit::toPyObject(std::move(result));
```
hm, good point. It was crashing so consistently due to other GIL bugs. I wonder why I didn't see it here.

salexspb commented on torch/csrc/perf_tools/ThroughputBenchmark-inl.h:
```diff
@@ -0,0 +1,143 @@
+  std::vector<std::thread> callers;
+
+  // Children need the GIL
+  AutoNoGIL no_gil_guard;
```
Well, then we can just say that AutoNoGIL is nothing in the c++ only world, should be a pretty simple change if / when we want to support c++ only mode.

dzhulgakov commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+    return script_module_.benchmark(config);
+  } else {
+    CHECK(module_.initialized());
+    LOG(WARNING) << "Starting benchmark on an nn.Module. This can be slow due "
```
if you really want warning - it's important to pass it to python's logger. I think TORCH_WARN does that

salexspb replied:
@dzhulgakov , not 100% sure what you mean here. I see that default warning handler is just cerr and there is additional data about file and line. What does passing to Python logger here mean and why is it needed?

dzhulgakov commented on torch/csrc/utils/throughput_benchmark-inl.h:
```diff
@@ -0,0 +1,142 @@
+  std::vector<std::thread> callers;
+
+  // Children need the GIL
+  AutoNoGIL no_gil_guard;
```
I think it'd be cleaner to remove it from here and assume that benchmar() is called without GIL. Then only methods explicitly specialized on ModelInput would be grabbing GIL (e.g. cloneInput). On python interface - you can just release GIL in pybind bindings.

It makes it really clean to think about GIL and also allows to call BenchmarkHelper directly from C++! (since we're really close to it- would be nice to do it too)

dzhulgakov commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+void ModuleBenchmark::runOnce(ModuleInput&& input) const {
+  CHECK(initialized_);
+  AutoGIL gil_guard;
+  // TODO: provide guarantees that compiler won't optimize this out
```
I think it won't - there's a lot happening inside forward.

dzhulgakov commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+  CHECK(initialized_);
+  AutoGIL gil_guard;
+  // TODO: provide guarantees that compiler won't optimize this out
+  model_.attr("forward")(*input.args, **input.kwargs);
```
btw, wouldn't model_(...) work?

zheng-xq commented on torch/csrc/utils/throughput_benchmark-inl.h:
```diff
@@ -0,0 +1,142 @@
+  std::vector<std::thread> callers;
+
+  // Children need the GIL
+  AutoNoGIL no_gil_guard;
```
Are the AutoNoGIL and "// Children need the GIL" still needed?

salexspb replied:
They are still needed. But as Dima pointed out, this indeed can be moved higher up in the stack

zheng-xq commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,187 @@
+ * is exposed as an object to python with the following read-only properties:
+ *
+ * stats.qps
+ * stats.num_forwards
```
Is this "num_iters"? Is "qps" still here?

salexspb replied:
thanks, nice catch

salexspb commented on torch/csrc/utils/throughput_benchmark-inl.h:
```diff
@@ -0,0 +1,142 @@
+      config.num_worker_threads == 1,
+      "Only parallelization by callers is supported");
+
+  // We pre-generate inputs here for each of the threads. This allows us to
```
I think user doesn't pay the cost of copy. User creates input somehow on the side. It is likely even slower than a copy to convert Thrift request to tensors. But it is not related to copy workload. So I prefer it this way.

salexspb commented on torch/csrc/utils/throughput_benchmark-inl.h:
```diff
@@ -0,0 +1,142 @@
+                  << ". Total number of initialized threads: " << initialized;
+        worker_main_cv.notify_one();
+        while (!start) {
+          LOG(INFO) << "Not starting this thread yet";
```
This won't hurt and it is not "production", but can be a bit too much logging though indeed. I would ask you to trust me cleaning this up before landing :) I will leave some logs and delete others.

apaszke replied:
Well we should never have any unconditional logs. We can have a `verbose` flag that lets people enable this, but by default no library functions should print anything.

salexspb replied:
I will remove extra things like this one, but having logs like starting the thread is useful IMO. This is not something which will run as a part of a bigger pipeline (like training a model or inference service), so it is reasonable to have some logs, though not this one specifically as XQ pointed about above.

salexspb commented on torch/csrc/utils/throughput_benchmark-inl.h:
```diff
@@ -0,0 +1,142 @@
+
+  TimePoint start_time;
+  {
+    std::unique_lock<std::mutex> lock(m);
```
Yep, makes sense, would be easier to reason about this.

dzhulgakov reviewed (approved):
Looks good! Double-check CI failures, otherwise is good to go!

salexspb commented:
oops, this one looks related, Windows build : https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/58850/console

dzhulgakov commented:
@bddppq pointed me to https://pybind11.readthedocs.io/en/stable/faq.html#someclass-declared-with-greater-visibility-than-the-type-of-its-field-someclass-member-wattributes - so it seems that pybind11 has this issue.

You can either declare your classes explicitly as `__attribute__((__visibility__("hidden")))` or we really should try to enable -fvisibility=hidden for the python extension (trying it now) - it was already enabled for C2's extension

dzhulgakov commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+  bool initialized_{false};
+};
+
+struct __attribute__((__visibility__("hidden"))) ModuleInput {
```
you need to ifdef it only on _WIN32 (feel free to add C10_HIDDEN to c10/macros/Export.h if you want)

salexspb replied:
ah, I missed this comment :( Thanks, Dima!

apaszke commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,87 @@
+    under an inference server like load. It can emulate multiple calling threads
+    to a single module provided. In the future we plan to enhance this component
+    to support inter and intra-op parallelism as well as multiple models
+    running in a single process.
```
1. No one cares if this is a C++ binding or not. Describe what this thing does in the docs, not the implementation details.
2. Future plans do not go into the docs. When you add the extra features (like intra-op parallelism), you'll update the docs. Leave this out for now.
3. Why is this specific to a server-like deployment? Measuring throughput of models doesn't require that, and it's a valid thing to do while performing research.

salexspb replied:
1. I disagree . This is a part of showing how we designed things and why. It is the next level comparing to docs which don't answer the "why" question
2. They might. We show future developers how to modify this code.
3. I think there is a misunderstanding here , ideally we discuss this in person if you are interested. Tl/Dr - one model many callers is pretty specific to deployment scenario where you have an inference service 
We might be interested in this on other scenarios, I agree. That's why I use "like" here

apaszke commented on torch/utils/throughput_benchmark.py:
```diff
@@ -0,0 +1,87 @@
+                num_iters = 1000,
+            )
+        >>> print("Avg latency (ms): {}".format(stats.latency_avg_ms))
+        >>> print("Number of iterations: {}".format(stats.num_iters))
```
Do we really believe that this heavy OOP API is the right way to go? Why can't the benchmark utility be something like this? It looks much simpler and fits it all in a single line.
```py
stats = torch.utils.bench.throughput(module, args=(x, y), kwargs=dict(a=1, b=2), num_calling_threads=4, ...)
```

apaszke replied:
Additionally, aren't we concerned that running the same model over and over on the same input data might cause artifacts in the benchmark due to the input data remaining in cache?

salexspb replied:
1. We need our current API to populate the dataset first 
2. We are not running over the same input . Depending on the number of inputs provided we will have more or less cache hits. Ideally you provide a dataset which doesn't fit your l3 before running benchmark function

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+ */
+struct BenchmarkExecutionStats {
+  float latency_avg_ms{-1};
+  int64_t num_iters{-1};
```
Would be nice to add docs for this struct (what does `latency_avg_ms` mean in this case)

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+  // to support some sort of multi-threaded forward calls. We may change this
+  // setting in the future to support different intra and inter op parallelizm
+  // which is not available in PyTorch yet
+  int num_worker_threads{1};
```
If you plan to add things later, add them later. There's no need to have dead code in the codebase. It's not like adding this field will break BC.

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+  py::args args;
+  py::kwargs kwargs;
+};
+typedef py::object ModuleOutput;
```
nit: prefer `using` instead of `typedef`

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+      : args(std::move(args)), kwargs(std::move(kwargs)) {}
+
+  // ModuleInput(py::args args, py::kwargs kwargs)
+  //     : args(std::move(args)), kwargs(std::move(kwargs)) {}
```
Unused code?

salexspb replied:
removed in the last version

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+  // would race with Python
+  void runOnce(Input&&) const;
+  // This method is to be used when calling from Python dirrectly
+  Output runOnce(py::args&&, py::kwargs&&) const;
```
Why is this something you might want to do with a benchmark helper though? I think we should focus on writing a tool that does one thing, but does it well.

salexspb replied:
IMO this is a reasonable thing to have for validation purposes.

apaszke commented on torch/csrc/utils/throughput_benchmark.h:
```diff
@@ -0,0 +1,181 @@
+  // This method to be used in benchmark() method
+  // Note that there is no result. This way we don't have to call this under GIL
+  // even when running in the nn.Module mode. Otherwise destructor of the result
+  // would race with Python
```
Um, this looks plain unsafe. You can't call into Python code without calling the GIL. That's UB and it's asking for trouble.

salexspb replied:
Every time I call into python code I call this under GIL

apaszke commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+    "to Python GIL.For proper inference simulation you might want to switch to "
+    "a ScriptModule instead");
+    return module_.benchmark(config);
+  }
```
Since everything else is templated, why isn't this class? Every single method has an `if (script_modue_.initialized())` so it's basically trying to do two completely different things.

salexspb replied:
This is so we can have one object  in the python bindings, this class hides the if statements from users

apaszke commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+void ScriptModuleBenchmark::runOnce(ScriptModuleInput&& input) const {
+  CHECK(initialized_);
+  // TODO: provide guarantees that compiler won't optimize this out
+  model_->get_method("forward").function()(std::move(input));
```
Well this can have side effects, so I don't think any compiler is even allowed to optimize this out.

salexspb replied:
I removed the comment in the latest version

apaszke commented on torch/csrc/utils/throughput_benchmark.cpp:
```diff
@@ -0,0 +1,129 @@
+template <>
+ScriptModuleInput cloneInput<ScriptModuleInput>(
+    const ScriptModuleInput& input) {
+  return input;
```
This doesn't really clone inputs, what's going on?

salexspb replied:
it does clone the input, it does the copy

apaszke reviewed (changes requested):
There are some things that we should clean up and improve before this is merged. There's one place that might be running without the GIL which is unsafe, `cloneInputs` doesn't clone inputs, etc.

salexspb commented:
Hey @apaszke , I am going to kindly ask follow up offline with me about this PR to make sure we don't have too much back and forth here while I consider your last set of comments as non blocking with all the respect to you. Kindly note  that this PR was standing out for quite some time. I am going to respectfully land this for now and if you would be able to show me some draw backs of the current approach, I will iterate over it later. I think I addressed the feedback from 4 persons on this PR already who have been kindly spending a lot of time iterating over this PR for several weeks while there was no single comment from you here before. In my view addressing comments from 4 reviewers who have been highly involved on this PR it is sufficient for merging. The only reason I didn't merge this yet is annoying visibility issues which I now fixed thanks to @dzhulgakov 's and @bddppq 's help. I am also going to be in NYC next week, not sure where you are based, but if you are there, happy to chat in person! We are also going to have some sessions there specifically designed to address extra back and forth on diffs which was a BIG pain point for people in AI Dev Platform. This PR is included in the list! :) Please also note that I went into quite a bit of compromises here already. In my view having support for nn.Module is not required for merging this, though I compromised on this and added it in the first version which complicated the  things quite a bit and slowed this down.

salexspb committed:
PyTorch ThroughputBenchmark (#20766)

Summary:
This is useful for measuring inference performance of your
models. This is a very basic benchmark for now. We don't support
batching on the benchmark side, no inter and intra op parallelizm is
supported yet, just caller based parallelizm.

Main phylosophy here is that user should be able to provide inputs
from python and just stack them within the benchmark. API should be
exactly the same as passing inputs to module.forward.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20766

Test Plan: Added a new unit test

Differential Revision: D15435461

Pulled By: salexspb

fbshipit-source-id: e8759c417faf6c52807d5a685eed48e023cd40e9

suo commented:
@salexspb this PR appears to have broken windows builds on master :(

soumith commented:
fyi, unlanding because it broke windows build.

salexspb commented:
@suo, @soumith, do you have logs showing the failures by any chance? There was a single failing job when I landed and it seemed unrelated to my change. Speaking of which before some fixes I made there were indeed Windows failures and I was hopping they are now addressed

suo commented:
@salexspb Here is one: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/59380//console

In general, you can look at https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master for master breakage status

salexspb commented:
Apparently I might've missed the error due to it being in "full logs" only. Will try to resend tonight

salexspb commented:
revert of the revert: https://github.com/pytorch/pytorch/pull/22185
